apps/test-moisture/src/main.cpp

```diff
@@ -7,35 +7,32 @@ DigitalOut led1(LED1);
 int main() {
   led1 = 0;  // start with LED off
 
-  uint16_t moisture = 0;  // initialize default variables
+  float moisture = 0;  // initialize default variables
```

Comment:
I recall this value being a 10bit uint when we tested it; any reason for this change? (maybe it was meant to be a floating-point number and I'm just an idiot idk)

Reply:
I changed it because I expect most return types to be at most Float precision when we want to create a general interface. The original function likely was supposed to be uinit16_t but I was unsure if this would impact its performance. If specific sensors require specific data types(In this case does moisture have to be a uint16_t or will it run fine?) I'll create more sets of "read" functions for different data types @ChristopherA057 Do you know if this will run alright or will the change in type affect performance?

Reply:
yeah we want the return type of the read function of the sensor base class to be float bc other implementations of the base class might require float precision. reasonably, this shouldnt break anything. it doesnt change any of the driver code rly. we can do all the calculations inside the functions with ints and then just cast to a float. floats are cheap bc of FPU

Reply:
i just learned that C++17 created something called std::variant. in this context, it basially means that the read() function of the base class can return a float or an int. the cherno's video abt it is pretty good. its basically a union but you trade memory for safety. i would say that this is a good place to use something like this. ultimately, it's up to you Felix if you think its a good idea. I think melvin might have good insight on this too


 https://www.youtube.com/watch?v=qCc_Vqg3hJk&ab_channel=TheCherno